### DISCUSSION TITLE
Interface supplied tests.

### DISCUSSION THREAD

**Discussion by fat-rooster** [dotnet/csharplang#7027] [Language Ideas]
I've been thinking a lot recently about what implementation and specification really mean, how to separate them, and it got me thinking about interfaces. There is not really a strong separation between specification and implementation in C#, and part of this is that the most powerful 'specification' concept (interfaces) is limited in how much it can specify.

On the other side of things, Unit tests are great. The reason they are great is that they are a clear way of expressing specification, and as expressive as the language itself. If you want to know what a piece of code is _supposed_ to do, you can look at the tests.

This is just a broad concept of an idea, so no idea what the feature would finally look like, but the concept is to consolidate specification features into interfaces, while types remain implementation focused. This means that tests should be defined on interfaces (and be as visible), and the language should have features to support this model.

Importantly, this enforces the idea that the tests should be public, even if the code that implements it is not.

This relates to the 'by' keyword idea in that that is intended to provide sample implementations. While this would not produce a full specification for an interface, it could be part of a system for automatically producing tests. In particular, if a function is expected to produce a specific answer, and a user implements an optimised version of that function, we can automatically generate tests from sample data and the canonical implementation.

Pretty new to this, so maybe missing something obvious. Thanks for your time!

**Comment by HaloFour** [comment]
This feels like a testing framework discussion more than a language discussion, writing common tests for any implementation of a given interface.  I don't think you'd need a canonical example of an implementation to accomplish that, you can use the tests themselves as the specification.

One example might be to have an abstract generic class as the base class of the unit tests for a given implementation.  That way the unit tests for the implementation inherit all of the core tests which enforce the specification of the underlying interface.

**Reply by fat-rooster** [reply]
Yeah, I should probably explain more why I think this is a language concern. [This comic](https://www.commitstrip.com/en/2016/08/25/a-very-comprehensive-and-precise-spec/?) is a starting point with the TLDR being that specifications are the future of code, interfaces are specification, and tests are a frequently used tool in how we write specification. If your coding workflow consists of feeding tests into an AI, and those tests are written in a 3rd party framework, are you really coding in C#?

A language can be thought of as expressing two distinct things; _how_ we do something, and _what_ we are doing. What objects _are_ vs what they _do_. Implementation and Specification. As we go forward (and AI tools in particular get better), languages that are better in terms of writing specifications that automated tooling can understand will probably win out*. There are a lot of features already available that make our implementations better and cleaner, but very few features that are designed to make the language more expressive in terms of specifications**. This looks short sighted to me.

An important sub part of that is separation between code dealing with implementations and code dealing with specifications. They should be kept distinct where possible, because they have different requirements**. The natural place for the dividing line is along interfaces. Interfaces are already a starting point for more general specifications, and used that way. When writing a function, we are encouraged to take interfaces as arguments because we care what they _do_, not what they are. When extending specification features, interfaces seem a good jumping off point.

The thing is that specification driven development isn't actually something new, it is just a generalisation of various techniques already in use that use 3rd party tools. Specifically, tests are an easily understood and frequently used tool in defining a specification. As a starting point for specification specific tooling it makes sense not to reinvent the wheel, and instead look at what people already do. What they do is test. The line of reasoning then goes that we focus our specification code around interfaces, and our specification code should include tests, hence tests on interfaces.

Before it comes up, default implementations do not pull in this direction. As a general rule, specification code should never hit the end user (exceptions exist, such as debug telemetry). Default implementations are provided by the downstream user, so are clearly implementation code. To drive this home we can ask: should default implementations be written with performance or readability as a larger concern? If performance is ever going to be a concern then you are almost certainly dealing with an implementation detail. I'm not knocking Default implementations, because they are awesomely powerful, but they are not specification code, so not really relevant to what I am talking about.

Traits are an example where default implementations are held high, and they definitely can implement them, but they have some deficiencies. Either you enforce a trait with a sealed default implementation (preventing implementers optimising where appropriate), or you leave it unsealed and implementers can get it wildly wrong. The language gives us no features to find a middle ground. Traits are either massively over, or under, specified. The default implementation can be regarded as canonical, but it is then pulled in different directions by performance and clarity concerns, as well as not being minimal**.

Canonical code is not 'needed' for a specification, but in the absence of any other explicit information about a piece of code they can be regarded as good starting point. They are however not minimal**. They give an example of what is sufficient to meet a specification, but the bar may be much lower. For example, the domain of the function may actually be quite restricted***. Behaviour outside the domain is undefined by the specification, but canonical code will do something. That 'something' is not part of the desired specification, despite being the behaviour of the canonical code. Implementers may find a better way of meeting the specification that does something else outside the domain, and that is completely fine. In reality when writing simple code quickly, writing a canonical example of an implementation is often just the simplest and fastest way of providing a specification, and we absolutely want to support that feature. In terms of backwards compatibility; we can produce a basic specification for any historic object by saying that it is the canonical example of it's own specification.

Probably the most important aspect of new canonical code is that it does not care about performance, so can be written for clarity. The unoptimized first draft of a function is often just the simplest possible expression of what it is supposed to do. Source control does a great job of preserving it for posterity and reference, but that loses it's utility as documentation. Code as documentation has a central issue arising from the fact that documentation cares about clear specifications, while we mostly write implementation. The clarity is often better in early versions of code. Specification code does not have this issue. Instead of letting source control simply replace the previous (working) version of the function and generate a new canonical specification, it would be preferable to instead have the new function still refer to the old function's specification (despite losing the old function itself), and auto generate tests for the new version (with the specifics up to the versioning engine). The net effect would be that the natural process of iteratively writing code would produce much better documentation without additional programmer work, because each step adds on the information of the previous one rather than discarding it. A reader who wants to understand what something does (and so looks to the specification because '_does_') would find themselves staring at the original clearer version of the function even if no documentation or detailed specification work had been done at all, and the actual final optimised implementation's source was not public.

Thanks for your time if you made it this far. I guess the title is a little bit prejudiced by my own beliefs about language design, and I am suggesting a limited feature that relies on a model that is not generally known, never mind accepted. Maybe a more general discussion about implementation/specification/configuration separation and whether it should be part of the design philosophy is warranted. 

*AI generated implementations together with theorem proving tools to show that the code meets a given specification would not be part of the language itself, but how we express the specification absolutely should be.

**There is a major distinction between implementation code and specification code: Implementation code wants to be expressive, and specification code wants to be minimal. Implementations have to supply all the detail, so implementations that are over specific are fine, as long as they include enough. Every detail must be supplied by an implementation, meaning that conflicts are a problem. Specifications can be under specific without problem (we can get away with none at all), and we do not want to over-specify. We can continue to add information to the specification code, and redundancy is not a problem. These two requirements ask for different things form the language.

***As a concrete example, consider `ATan2(x,y)`. This has a well defined return when either x and y are non zero, but what should it do if both are zero (also hard to express cleanly with types)? There is no natural answer to that. We can coerce all implementations to produce a certain value, but that may have a price on some of them, including any better future implementations. It makes more sense to say that the function has no defined behaviour for this point. Implementations are free to do what is convenient, while a separate specification is required to expresses this fact to users. A canonical implementation could sort of show this, by throwing, but that doesn't tell users whether throwing is expressing undefined behaviour or that the defined behaviour is to throw.
